### PR TITLE
VB-574: Main contact - form fixes and pre-populate from session

### PR DESCRIPTION
**Pull request by tpmcgowan**

### CODE REVIEW

File: server/routes/visitors.ts [hunk 1]
```diff
@@ -290,11 +295,22 @@ export default function routes(
       res,
     })
 
-    const { offenderNo } = req.session.visitSessionData.prisoner
+    const formValues = (req.flash('formValues')?.[0] as unknown as Record<string, string | string[]>) || {}
```
Comment (psoleckimoj) [inline server/routes/visitors.ts:298]: If this is used everywhere, maybe stick in a function?

File: server/routes/visitors.ts [hunk 2]
```diff
@@ -290,11 +295,22 @@ export default function routes(
...
+
+    if (!Object.keys(formValues).length && req.session.visitSessionData.mainContact) {
```
Comment (psoleckimoj) [inline server/routes/visitors.ts:300]: Maybe assign `req.session.visitSessionData` to a variable?

Comment (psoleckimoj) [reply server/routes/visitors.ts:300]: ..to make it easier to read.

**Review by psoleckimoj — DISMISSED** [review]
All looks good with a couple of minor comments. Also you'll have to explain why the data structure needed changing as I can't see the reason from the code 🤷‍♂️

**Review by psoleckimoj — APPROVED** [review]
LGTM!

File: server/routes/visitors.ts (hunk 1 above)
Comment (tpmcgowan) [reply server/routes/visitors.ts]: Good suggestion - have done this 👍

File: server/routes/visitors.ts (hunk 2 above)
Comment (tpmcgowan) [reply server/routes/visitors.ts]: Done - in fact as it was repeated a lot I got carried away and did it throughout this file and the utils file.

**Review by psoleckimoj — APPROVED** [review]
LGTM!

**Comment by tpmcgowan** [comment]
@psoleckimoj about the data structure change, I'm assuming you were referring to e.g. `adultVisitors: VisitorListItem[]` becoming `adultVisitors: { adults: VisitorListItem[] }`? This was because when I tried to pass `adultVisitors` via `appWithAllRoutes(...)` I found I couldn't test what had happened as it got overwritten in the code. Passing an object with this value inside seemed to let me inspect it afterwards to see what had happened.

**Pull request merged**